File: Grind169/Week1/733-Flood-Fill/solution.py
from typing import List


class Solution:
    def floodFill(
        self, image: List[List[int]], sr: int, sc: int, color: int
    ) -> List[List[int]]:
        # Use BFS

        origin_color = image[sr][sc]
        m = len(image)
        n = len(image[0])

        if origin_color != color:
            stack = [(sr, sc)]

            while stack:
                row, col = stack.pop(0)

                image[row][col] = color

                for x, y in [
                    (row - 1, col),
                    (row + 1, col),
                    (row, col - 1),
                    (row, col + 1),
                ]:
                    if 0 <= x < m and 0 <= y < n and image[x][y] == origin_color:
                        stack.append((x, y))

        return image

File: Grind169/Week1/733-Flood-Fill/solution.cpp
#include <vector>

using namespace std;

class Solution {
public:
    void dfs(vector<vector<int>>& image,int row,int col,int start_color,int color){
        if (row<0 || row>=image.size() || col <0 || col>=image[0].size() || image[row][col]==color || image[row][col]!=start_color) {
            return;
        }

        image[row][col]=color;

        dfs(image,row-1,col,start_color,color);
        dfs(image,row+1,col,start_color,color);
        dfs(image,row,col-1,start_color,color);
        dfs(image,row,col+1,start_color,color);
    }


    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        // Use DFS
        int start_color=image[sr][sc];

        dfs(image,sr,sc,start_color,color);

        return image;

    }
};